core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java

```diff
@@ -180,8 +180,10 @@ protected void generateTypeString(StringBuilder sb, boolean withDetail) {
     boolean printScale = scale != SCALE_NOT_SPECIFIED;
 
     // for the digest, print the precision when defaulted,
-    // since (for instance) TIME is equivalent to TIME(0).
-    if (withDetail) {
+    // but for TIME or TIMESTAMP, although (for instance)
+    // TIME is equivalent to TIME(0), the digest of TIME
+    // and TIME(0) should be different literally.
+    if (withDetail && !SqlTypeName.DATETIME_TYPES.contains(typeName)) {
       // -1 means there is no default value for precision
```

Comment:
Use `SqlTypeUtil.isDatetime`

Reply:
`SqlTypeUtil.isDatetime` accepts `RelDataType`, but here the type is `SqlTypeName`, so I was thinking it is not able to change. 
`RelDataType --contains--> SqlTypeName --contains--> SqlTypeFamily`